### ISSUE TITLE
cdms2 and library updates

### ISSUE THREAD

**Issue by mauzey1** [PCMDI/cmor#629]
There are currently pending pull requests to the CMOR conda-forge feedstock that update the libnetcdf (https://github.com/conda-forge/cmor-feedstock/pull/41) and hdf5 (https://github.com/conda-forge/cmor-feedstock/pull/40) libraries used in the stable release builds.

The latest version of CDMS2 from conda-forge is not compatible with either of the updated libraries.  If we want to continue supporting the use of CDMS2 with CMOR, then we need an updated build of CDMS2 that uses the newer versions of libnetcdf and hdf5.  The alternative would be to remove CMDS2 support from CMOR.

@durack1 @taylor13 I would like your thoughts on this.  Also, please ping anyone else that this issue would concern.

**Comment by durack1** [comment]
@mauzey1 thanks for the update. My preference would be to keep these in sync, but obviously that would require cdms2 to be updated.

@tomvothecoder @jasonb5 what are the chances that cdms2 v3.1.6 could be released with dependency updates? I think that E3SM still depends on cdms2 and CMOR to write files for ESGF publication (@sterlingbaldwin)

Ping @pochedls @gleckler1

**Comment by durack1** [comment]
@chengzhuzhang pinging you too!

**Comment by taylor13** [comment]
As I understand it, cdms2 is needed to run the test codes, but not for normal execution of CMOR.  Is this true?

**Comment by mauzey1** [comment]
@taylor13 There are parts of CMOR that support reading the values in the format that CDMS2 stores them as.  This is meant to support reading values from NetCDF files opened using CDMS2, which some of the tests use.

**Comment by taylor13** [comment]
I wonder if any external users rely on CMOR to read in data (besides when running the tests).  It seems like reading data is pretty separate from writing data, so couldn't the two packages be offered independently?  What are the arguments for combining them?

**Comment by durack1** [comment]
@taylor13 they are offered as completely separated packages (see https://anaconda.org/conda-forge/cdms2 and https://anaconda.org/conda-forge/cmor), but the conflicts with their ***dependencies*** as they are separate packages is the issue at question here

**Comment by mauzey1** [comment]
CMOR itself doesn't rely on CDMS2 but can be used in conjunction with CMOR.  CDMS2 can be used for opening NetCDF files instead of using NetCDF4 in Python.  However, there will be package dependency conflicts as newer versions of CMOR get released while CDMS2 remains stagnant.

**Comment by taylor13** [comment]
As you noted above,

"The latest version of CDMS2 from conda-forge is not compatible with either of the updated libraries. If we want to continue supporting the use of CDMS2 with CMOR, then we need an updated build of CDMS2 that uses the newer versions of libnetcdf and hdf5. The alternative would be to remove CMDS2 support from CMOR."

If we go with the second option and we "remove CDMS2 support from CMOR," won't we be saying in effect "you can't use CDMS2 at the same time you use CMOR because this will cause dependency conflicts (because CMOR and CDMS2 require different versions of the supporting software)"  Is that correct?

**Comment by mauzey1** [comment]
@taylor13 Yes, CDMS2 and CMOR can't be installed in the same environment since they will rely on different versions of the same dependent libraries.

**Comment by chengzhuzhang** [comment]
Thank you for the discussion! Only speaking for the tool E3SM is using to cmorize model output (e3sm_to_cmip), It started to use xarray to read in data. There might be other external users rely on having both cdms2 and cmor in the same environment? 
And question for @jasonb5 do you think it is worthwhile to build cdms2 with the newer versions of libnetcdf and hdf5?

**Comment by taylor13** [comment]
I learned from P. Gleckler that obs4MIPs relies on simultaneously using CMOR and cdms2.  He said his expectation is that cdms2 will be supported for perhaps another year or two after which xcdat will be mature enough that we should be able to transition away from cdms2.  If that is the case, then I think we should try to make sure cmor and cdms2 remain compatible for at least a year.  At that time support for cdms2 will go away, and anyone will need to adapt to using other options for reading in data before writing with cmor.  Does this make sense?
I think @jasonb5 will need to be consulted, since as I understand it, he is the only person who can maintain cdms2.

**Comment by jasonb5** [comment]
@chengzhuzhang I think it's worthwhile to update cdms2 with the new versions of libnetcdf and hdf5.

As long as the new libraries don't break cdms2 I'll probably just bump the build number rather than release a new version.

**Comment by chengzhuzhang** [comment]
@jasonb5 Thank you. If we don't hear objection from this post, let's try update cdms2 with new libnetcdf and hdf5..
I suppose there is a test suite in place to check if the new libraries would break cdms2?
Pinging @xylar too, Xylar do you expect updating cdms2 with new libnetcdf and hdf5 would have some side effect in the upcoming `e3sm-unified` release?

**Comment by durack1** [comment]
@jasonb5 thanks for this, so I assume that we'd see a `build = 10` appear in https://anaconda.org/conda-forge/cdms2/files - so something like `osx-64/cdms2-3.1.5-py39hc54994c_10.tar.bz2`?

**Comment by xylar** [comment]
@chengzhuzhang and @jasonb5, yes, we very much need to update cdms2 to these latest versions (which presumably means updating the dependencies).  Otherwise, E3SM-Unified likely won't work.

**Comment by xylar** [comment]
I will try to figure out where the migration bot is stuck that `cdms2` hasn't been migrated to these library versions yet.

**Comment by xylar** [comment]
It was stuck here:
https://github.com/conda-forge/libcf-feedstock/pull/46
I just merged that.  The bot should generate a `cdms2` PR soon and we should be good.

In the meantime, I would recommend merging the `cmor` PRs.  It will still be possible to install `cdms2` and `cmor` with older build numbers so there will be no harmful side effects in doing so.  New versions of `cmor` after those PRs would, however, not be compatible with `cdms2` until we get the `libnetcdf` version updated.

**Comment by durack1** [comment]
@xylar thanks for chiming in, I assume that you folks are using CMOR to write E3SM data, or is my assumption wrong?

**Comment by xylar** [comment]
@chengzhuzhang can confirm but I believe our `e3sm_to_cmip` project uses `cmor`.  So we use it for post-processing.

**Comment by chengzhuzhang** [comment]
@durack1 Xylar is correct. We use CMOR to reformat E3SM data to be CMIP compliant. I was wrong earlier, we do need `cmor` and `cdms2` to live together,  in `e3sm-unified`.

**Comment by durack1** [comment]
@chengzhuzhang glad to clarify that, it makes management of dependencies infinitely easier if we're not diverging across projects. cdms2 3.1.6 (or build 10) here we come!

**Comment by xylar** [comment]
@chengzhuzhang, we do have both `cdms2` and `e3sm_to_cmip` in `e3sm-unified`.  But merging the two PRs to update the version of `libntecdf` and `hdf5` will not make that any easier or harder for now.  So I would recommend to `cmor` folks to keep pins updated as `conda-forge` migrates them and not to worry about what `cdms2` is doing.  We can make sure `cdms2` gets migrated as well in a reasonable amount of time.  The thing is that the conda solver will find an older compatible build if the newest build has incompatibilities.  So it's not like a new build with a new `libnetcdf` or `hdf5` replaces the older build once and for all.

**Comment by xylar** [comment]
@mauzey1 and others,

> The latest version of CDMS2 from conda-forge is not compatible with either of the updated libraries. If we want to continue supporting the use of CDMS2 with CMOR, then we need an updated build of CDMS2 that uses the newer versions of libnetcdf and hdf5. The alternative would be to remove CMDS2 support from CMOR.

I'm confused by this original paragraph.  On conda-forge, `cdms2` is not a dependency of `cmor` in its recipe (either directly or as a dependency of a dependency). Otherwise, the bot would not have created the two migration PRs that are mentioned at the top.  Is that an error and should `cmor` be a dependency in the recipe on conda-forge?

**Comment by xylar** [comment]
Sorry, I read the whole thread more carefully and I now understand.

There was an argument earlier in the thread that it might be best to hold off on https://github.com/conda-forge/cmor-feedstock/pull/41 and https://github.com/conda-forge/cmor-feedstock/pull/40 because then future releases of `cmor` would remain compatible with `cdms2`, which has not yet migrated.  The problem with this is that most other packages in the conda-forge scientific packages stack *are* migrating to these new versions so the number of new packages that you won't be able to install because you *didn't* migrate will soon be a much bigger problem. In my experience as a maintainer of ~100 conda-forge packages, it's nearly always better to migrate right away when the bot asks you to do so.  If incompatibilities emerge, it is nearly always preferable to try to get the stragglers migrated than to hold up the migration.  The more packages that get released while migration is on hold, the more likely that folks like me who maintain some pretty massive and fairly heavily constrained environments can't find a solution.

So, again, I would encourage you to merge those 2 PRs and we will do our best to make sure `cdms2` and its dependencies also keep up with migration.

**Comment by xylar** [comment]
I'm manually doing the `cdms2` migration here: https://github.com/conda-forge/cdms2-feedstock/pull/73 in case you want to follow along.

**Comment by xylar** [comment]
Build 10 of `cdms2` should now be available with the latest `libnetcdf` pin on conda-forge.

**Comment by jasonb5** [comment]
@xylar Thanks for taking care of https://github.com/conda-forge/cdms2-feedstock/pull/73

**Comment by mauzey1** [comment]
The CDMS2 and CMOR conda-forge builds have been updated.  Closing this issue.

**Comment by durack1** [comment]
@mauzey1 excellent! Thanks too @xylar for pushing `cdms2` along

**Comment by xylar** [comment]
@mauzey1, thanks for raising this issue in the first place.  I usually try to keep up with migrations like these but I lost track of the `libcdms` one somehow.  Glad we're back in sync!
